### ISSUE TITLE
Desktop Agent Bridging Discussion group 8th June 2022

### ISSUE THREAD

**Issue by kriswest** [finos/FDC3#741]
## Group overview
Discussion group for developing proposals for producing bridges between FDC3 implementations (aka Desktop Agents), allowing applications running on one Desktop Agent, to integrate with FDC3 applications running on a second Desktop Agent for the same user.

The interop between applications running on different Desktop Agents aka Platforms would ideally cover
- Channels
- Raising and resolving Intents
- Launching Applications

### Relevant issue tags
- Current open issues that relate to the above concepts with the label: 
  [![image](https://user-images.githubusercontent.com/1701764/144856698-e49821ab-9c8e-4daf-99a1-23f4774c1fcd.png)](https://github.com/finos/FDC3/labels/Desktop%20Agent%20Bridging)
- The original issue is now a github discussion where content can be added and this issue discussed: 
#544

## Meeting Date
Wednesday 08 June 2022 - 9am EST / 2pm BST

### WebEx info
- Meeting Link: [https://finos.webex.com/finos/j.php?MTID=m55d4aa79e148c948bf53163690e74fee](https://finos.webex.com/finos/j.php?MTID=m55d4aa79e148c948bf53163690e74fee)
- Meeting number: 2558 920 8729

*More ways to join*
- Join by video system:
   - Dial 25589208729@finos.webex.com
   - You can also dial 173.243.2.68 and enter your meeting number
- Join by phone
   - +1-415-655-0003 US Toll
   - +44-20319-88141 UK Toll
- Access code: 2558 920 8729

## Meeting notices
- FINOS **Project leads** are responsible for observing the FINOS guidelines for [running project meetings](https://github.com/finos/community/blob/master/governance/Meeting-Procedures.md#run-the-meeting). Project maintainers can find additional resources in the [FINOS Maintainers Cheatsheet](https://odp.finos.org/docs/finos-maintainers-cheatsheet/).

- **All participants** in FINOS project meetings are subject to the [LF Antitrust Policy](https://www.linuxfoundation.org/antitrust-policy/), the [FINOS Community Code of Conduct](https://github.com/finos/community/blob/master/governance/Code-of-Conduct.md) and all other [FINOS policies](https://github.com/finos/community/tree/master/governance#policies).

- FINOS meetings involve participation by industry competitors, and it is the intention of FINOS and the Linux Foundation to conduct all of its activities in accordance with applicable antitrust and competition laws. It is therefore extremely important that attendees adhere to meeting agendas, and be aware of, and not participate in, any activities that are prohibited under applicable US state, federal or foreign antitrust and competition laws. Please contact legal@finos.org with any questions.

- FINOS project meetings may be recorded for use solely by the FINOS team for administration purposes. In very limited instances, and with explicit approval, recordings may be made more widely available.
  
- A Discussion Group has no direct decision-making power regarding the FDC3 standard - rather it is intended that anything they propose or work on will result in proposals (via Github issues and PRs) for the Standards Working Group participants to consider and vote on for inclusion in the standard. As such, participation in a Discussion group is not required for contributing to any particular issue or FDC 2.0 as a whole.

## Agenda
- [x] Convene & roll call, review meeting notices (5mins)
- [x] Review action items from previous meeting (5mins)
  - #692 
  - FDC3 2.0 timeline and thoughts on 2.1
  - Current proposal PR: #634
    - Preview: https://github.com/finos/FDC3/blob/544-Desktop-Agent-Bridging-Proposal/docs/api-bridging/spec.md
- Review & discuss open questions:
  - [x] Handshake, authentication and naming protocol (15mins)
    - Should auth be optional? E.g. triggered by a challenge?
    - See notes in proposal
  - [x] How do we resolve differing state on channels (app/user/private) when a new DA joins the bridge? (15mins)
    - See notes in proposal
  - [x] Who collates responses to queries (e.g. findIntent, raiseIntent, findInstances etc.)? Bridge or requesting DA? (10mins)
    - How to handle slow responding DAs? Should there be a recommended timeout?
    - Should an agent that is repeatedly timing out be disconnected?
    - Should other agents report to users on connect/disconnect events?
  - [x] Select recommended port range for bridge (5mins)
- [x] AOB & Adjourn (5mins)

## Minutes

- **Handshake, authentication and naming protocol** 
  - Handshake
    - @robmoffat - Suggested [STOMP](http://jmesnil.net/stomp-websocket/doc/) messaging protocol - [STOMP Websocket](https://stomp-js.github.io/stomp-websocket/)
    - @kriswest  - Mentioned other websocket frameworks such as [Socket.IO](https://socket.io/)
    - No strong opinions from attendees on the handshake messaging format
      - Update proposal with a suitable set of handshake object formats (Bridge + DA)
  - Authenticating connections
    - Several options presented
      - @timjenkel: 
        - the options fall into two groups: authenticating that a DA is a particular DA and authenticating the user using that DA
        - expressed a preference for JWT but thinks there are other use-cases in which other mechanisms might be better suited
      - There was consensus that we should
        - remove user auth for now - it was acknowledged that that use-cases might exist in the future and can be brought in into the proposal but for now will be considered out of scope.
        - use JWT for DA auth to bridge
          - may be combined with sending of implementation details (JWT signs a block of base64 information)
  - No objections with the current proposal in regards to the Connection Protocol
    - Recommend that DA's notify users of connect/disconnect of _other_ agents, but not required for compliance

- **How do we resolve differing state on channels (app/user/private) when a new DA joins the bridge?**
  - @kriswest Enumerated the different problems/challenges on app joining channels and context synchronization and ran through the possible solutions (see Open Questions section of proposal)
  - There was consensus that we should:
    - Adding an init flow to the connection protocol based on 'First to join' channel context caching (i.e. existing context from DA's already connected to the bridge will overwrite context from an agent that is connecting, however context for channels that did not already have it will have the new agent's context merged in/broadcast)
    - Bridge will have to share known channels and current state on join
    - Additional state will need to be kept in the bridge (current context of each known channel)
    - Ensure DAs join the bridge serially to enable context merges

- **Who collates responses to queries (e.g. findIntent, raiseIntent, findInstances etc.)? Bridge or requesting DA?**
  - There was consensus that the Bridge should collate the DA responses and send on to the requesting DA. 
    - This adds complexity to the bridge implementations but simplifies all DA implementations.
  - Bridge should also implement timeout for waiting on DA responses. 
    - Recommended a timeout at most 2500ms/3000ms since it is assumed that the communication will on the same machine. 
    - The bridge MUST allow for timeout configuration
      - Customisation more likely needed in multi-machine use-cases?

- **Select recommended port range for bridge**
  - No recommendations made during meeting.
  - Investigation needed to come up with recommended PORT range, select a block.
    - registered ports: https://tools.ietf.org/id/draft-cotton-tsvwg-iana-ports-00.html#userports
      - forms for requesting port number registrations on the IANA web site [[USRFORM]](https://tools.ietf.org/id/draft-cotton-tsvwg-iana-ports-00.html#USRFORM)
      - registry: http://www.iana.org/assignments/port-numbers (filter for unassigned ranges)
    - dynamic range: https://tools.ietf.org/id/draft-cotton-tsvwg-iana-ports-00.html#privateports
      - probably better to use an unassigned registered range

## Action Items
- [x] @tpina @kriswest incorporate feedback into and continue development of handshake, auth and connection protocol
  - Add Handshake message formats
  - Describe auth scheme and required config
- [x] @tpina @kriswest Add a channel/context sync step to connection protocol as agreed in minutes
- [ ] @tpina @kriswest Clarify collection of query responses as responsibility of bridge and start working into proposals for individual message exchanges.
- [ ] @tpina @kriswest Select a port range from IANA registry  

## Untracked attendees

| Full name | Affiliation | GitHub username |
|-----------|-------------|-----------------|
|           |             |                 |

**Comment by kriswest** [comment]
Kris / Cosaic

**Comment by timjenkel** [comment]
Tim Jenkel / Wellington
I will need to drop after first half

**Comment by mistryvinay** [comment]
Vinay / Symphony

**Comment by WatsonCIQ** [comment]
Chris Watson / Cosiac

**Comment by sarahstone-cosaic** [comment]
Sarah Stone / Cosaic
